src/lib/zkapp.ts

```diff
@@ -370,6 +450,134 @@ export class SmartContract {
   }
 }
 
+type StateUpdate<State, Update> = {
+  state: AsFieldElements<State>;
+  update: AsFieldElements<Update>;
+  apply(state: State, update: Update): State;
+};
+
+type StateUpdateReturn<State, Update> = {
+  emit: (update: Update) => void;
+  applyUpdates(
+    stateAndUpdates: {
+      state: State;
+      stateHash: Field;
+      updates: Update[][];
+    },
+    advancedOptions?: { maxTransactionsWithUpdates?: number }
+  ): {
+    state: State;
+    stateHash: Field;
+  };
+};
+
+function getStateUpdate<S, U>(
+  contract: SmartContract
+): StateUpdateReturn<S, U> {
+  let stateUpdate: StateUpdate<S, U> = ((contract as any)._ ??= {}).stateUpdate;
+  if (stateUpdate === undefined)
+    throw Error(
+      'You are trying to use a state update without having declared its type.\n' +
+        `Make sure to add a property \`stateUpdate\` on ${contract.constructor.name}, for example:
+class ${contract.constructor.name} extends SmartContract {
+  stateUpdate = {
+    state: Field,
+    update: Field,
+    apply(state: Field, update: Field) {
+      return state.add(update);
+    }
+  }
+}`
+    );
+  return {
+    emit: (update: U) => {
+      let party = contract.self;
+      let eventFields = stateUpdate.update.toFields(update);
+      party.body.sequenceEvents = Events.pushEvent(
+        party.body.sequenceEvents,
+        eventFields
+      );
+      if (mainContext?.otherContext?.numberOfSequenceEvents !== undefined) {
+        mainContext.otherContext.numberOfSequenceEvents += 1;
+      }
+    },
+
+    applyUpdates(
+      {
+        state,
+        stateHash,
+        updates: updateLists,
+      }: {
+        state: S;
+        stateHash: Field;
+        updates: U[][];
+      },
+      { maxTransactionsWithUpdates = 32 } = {}
+    ): { state: S; stateHash: Field } {
+      if (updateLists.length > maxTransactionsWithUpdates) {
+        throw Error(
+          `stateUpdate.applyUpdates: Exceeded the maximum number of lists of updates, ${maxTransactionsWithUpdates}.
+Use the optional \`maxTransactionsWithUpdates\` argument to increase this number.`
+        );
+      }
+      let methodData = contract.analyzeMethods();
+      let possibleUpdatesPerTransaction = [
```

Comment:
one of the magical things this function does is it knows how many events each of the smart contract methods emits, so it can perform computations for each of those numbers-of-events and then pick the actual one.
to enable that, I added `SmartContract.analyzeMethods` in this file and `synthesizeMethodArguments` in proof_systems.ts